clients/command-amqp/src/main/java/org/eclipse/hono/client/command/amqp/ProtonBasedCommandRouterClient.java

```diff
@@ -52,8 +62,34 @@
  */
 public class ProtonBasedCommandRouterClient extends AbstractRequestResponseServiceClient<JsonObject, RequestResponseResult<JsonObject>> implements CommandRouterClient {
 
+    /**
+     * Interval in "set last known gateway" requests are done.
+     */
+    static final long SET_LAST_KNOWN_GATEWAY_UPDATE_INTERVAL_MILLIS = 400L;
+    /**
+     * Maximum number of entries in a "set last known gateway" batch request.
+     */
+    static final int SET_LAST_KNOWN_GATEWAY_UPDATE_MAX_ENTRIES = 100;
+    /**
+     * The maximum number of "set last known gateway" requests triggered in one go. Subsequent requests will only be
+     * sent after these requests are finished.
+     */
+    static final int SET_LAST_KNOWN_GATEWAY_UPDATE_MAX_PARALLEL_REQ = 50;
+
     private static final Logger LOG = LoggerFactory.getLogger(ProtonBasedCommandRouterClient.class);
 
+    /**
+     * Key is a Pair with tenant and device identifier, value is the gateway identifier.
+     */
+    private final LinkedHashMap<Pair<String, String>, String> lastKnownGatewaysWorkQueue = new LinkedHashMap<>();
```

Comment:
Is there a particular reason why you are using the concrete class as the field's type here?

Reply:
To emphasize that it's a map where the iteration order is defined by the insertion order, hence also the `queue` part of the variable name.